plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/CabalAdd.hs

```diff
@@ -0,0 +1,160 @@
+{-# LANGUAGE ExplicitNamespaces    #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# LANGUAGE RecordWildCards       #-}
+
+module Ide.Plugin.Cabal.CabalAdd
+(  findResponsibleCabalFile
+ , missingDependenciesAction
+ , missingDependenciesSuggestion
+ , hiddenPackageAction
+ , cabalAddCommand
+ , command
+)
+where
+
+import           Control.Monad                          (void)
+import           Control.Monad.IO.Class                 (liftIO)
+import           Data.String                            (IsString)
+import qualified Data.Text                              as T
+import           Development.IDE                        (IdeState)
+import           Distribution.PackageDescription.Quirks (patchQuirks)
+import           Ide.PluginUtils                        (mkLspCommand)
+import           Ide.Types                              (CommandFunction,
+                                                         CommandId (CommandId),
+                                                         PluginId)
+import           Language.LSP.Protocol.Types            (CodeAction (CodeAction),
+                                                         CodeActionDisabled (CodeActionDisabled),
+                                                         CodeActionKind (CodeActionKind_QuickFix),
+                                                         Diagnostic (..),
+                                                         Null (Null), Uri (..),
+                                                         type (|?) (InR))
+import           System.Directory                       (doesFileExist,
+                                                         listDirectory)
+
+import           Data.Aeson.Types                       (FromJSON, ToJSON,
+                                                         toJSON)
+import           Data.ByteString                        (ByteString)
+import qualified Data.ByteString.Char8                  as B
+import           Data.List.NonEmpty                     (NonEmpty (..),
+                                                         fromList)
+import           Distribution.Client.Add                as Add
+import           Distribution.Compat.Prelude            (Generic)
+import           Distribution.PackageDescription        (packageDescription,
+                                                         specVersion)
+import           System.FilePath                        (dropFileName,
+                                                         splitPath,
+                                                         takeExtension, (</>))
+import           Text.Regex.TDFA
+
+
+-- | Given a path to a haskell file, finds all cabal files paths
+--   sorted from the closest to the farthest.
+--   Gives all found paths all the way to the root directory.
+findResponsibleCabalFile :: FilePath -> IO [FilePath]
+findResponsibleCabalFile uriPath = do
+  contents <- mapM listDirectory allDirPaths
```

Comment:
File IO is expensive, I'd suggest against travesing everything up to the very top. You can spray it generously with `unsafeInterleveIO`, or rewrite more explicitly with checking each level before proceeding to scan an upper one.

Reply:
I also wonder if we don't know this already. Do we not know what component a file belongs to, and thus hopefully what cabal file is relevant? At the very least we could compute it with a Rule.